### PR TITLE
Histogram functionality

### PR DESCRIPTION
**Pull request by MarsFlyer**
As discussed at Homecamp 4, I've added Histogram functionality.
The histogram shows the energy (watt hours) by power (watts) which allows analysis of what types of devices (low / medium / high) are using the most energy.
1. Adds a new Histogram feed type.
       Note this needs an additional column in the table. I've set this as process type 16.
2. Display a Histogram chart, this can be shown for any date range (year, month & day).
       This is set as view type 5. 

To do in the future:
1. Provide a right click zoom rather than using the checkbox.
2. Allow comparison of different dates (this will need normalisation for the number of days).
3. Highlight the low/medium/high dividing wattages (the watts where 33% & 66% of the energy has been used).

### CODE REVIEW

**Comment by MarsFlyer** [comment]
An image of the chart and checkbox (anyone know how to handle right clicks in flot?)
histogram

File: Models/feed_model.php [hunk 1]
```diff
@@ -12,8 +12,9 @@
   //----------------------------------------------------------------------------------------------------------------------------------------------------------------
   // Creates a feed entry and relates the feed to the user
   //----------------------------------------------------------------------------------------------------------------------------------------------------------------
-  function create_feed($userid,$name)
+  function create_feed($userid,$name,$processid)
   {
+    // *    TODO: Don't these need to handle the userid??
```
Comment (TrystanLea) [inline Models/feed_model.php:17]: User id's are handled by the feed_relation table

**Comment by TrystanLea** [comment]
Hello Paul, Thanks a great deal for this very large feature commit! Will have a good study of it today. Thanks a lot!

**Comment by MarsFlyer** [comment]
Actually the issue is with duplicate feed names - each user may want a feed with the same name, but the next line prevents duplicates across the whole database:
    $result = db_query("SELECT id FROM feeds WHERE name='$name'");              // Select the same feed to find the auto assigned id
(actually the $ido = db_insert_id(); is used) throughout.)

I tried creating a feed with a duplicate name and it created the input.processList entry, but none of the feed tables.

P.S. I'm getting this error message when editing feeds, it could be something I've changed:  
Warning: Cannot modify header information - headers already sent by (output started at C:\Program\xampp\htdocs\emoncms3\Controllers\feed_controller.php:152) in C:\Program\xampp\htdocs\emoncms3\Controllers\feed_controller.php on line 65

File: Includes/db.php [hunk 2]
```diff
@@ -42,12 +42,16 @@ function db_connect()
 
   function db_query($query)
   {
-    return $GLOBALS['mysqli']->query($query);
+    $ret = $GLOBALS['mysqli']->query($query);
+    if ($ret == false) {echo $GLOBALS['mysqli']->error;}
```
Comment (TrystanLea) [inline Includes/db.php:46]: great idea!

**Comment by TrystanLea** [comment]
A yes! I guess it did not make much difference as the $ido was used but the lines 

$result = db_query("SELECT id FROM feeds WHERE name='$name'");  
if ($result) {

could be replaced with:

if ($ido>0) {

Thanks for spotting that

"I tried creating a feed with a duplicate name and it created the input.processList entry, but none of the feed tables."

A user can only have unique feed names, while multiple users can have the same feed names. 
A process list is created as it allows a user to point an input at an existing feed. So lets say you had an input called power1 and linking to a feed called power but then you want to remap and link input power 2 to power and delete the old input, it makes that possible. Its something I seem to need to do quite often as input names change

**Comment by TrystanLea** [comment]
Does the error appear when editing: changing feed name, tag? I cant seem to replicate here?

**Comment by TrystanLea** [comment]
I'm thinking about how the histogram feeds are handled and am wondering whether it would be better to separate it out somehow from the normal feed handling functions. Maybe have something like:
create_histogram_feed($userid,$name)
and
get_histogram_feed_data($feedid,$start,$end,$resolution)
Maybe emoncms needs a way to accommodate different feed types better, as it is there is the issue that kwh/d feeds cannot be visualised by the realtime or rawdata visualisation and instantaneous data when attempted to visualise with kwhd vis causes the whole thing to crash as it attempts to load all the feed data.

Then with histogram feeds they cannot be visualised separately from the kwhd feed? the kwhdzoomer vis in general depend on a power and kwhd feed and multigraph can be used with any number of feeds.

I wonder what the best way forward is.

**Comment by TrystanLea** [comment]
I have created a emoncms3 dev branch, could you do a pull request to the dev branch and we can work on it there?

**Comment by MarsFlyer** [comment]
I originally put the histogram code in the original calls to just get it working as the model view controller approach can be difficult to navigate until you get your eye in. I agree separate calls would be better.
The modify header happens when I rename the feed - ignore it for the moment.
I've done the pull request into dev.

**Comment by TrystanLea** [comment]
Thanks a lot I've merged it there I will have another look at it later today

**Comment by TrystanLea** [comment]
Hello Paul, I have made a few changes: 
- The histogram vis is now in a different folder so we have both: kwhd_histogram and kWhdZoomer
- I have moved the point at which the number of data fields in the feed table is declared to the process_model list.
- This then via the process controller calls one feed creation function with an argument for number of fields required. This could allow for the number of data-fields to be extended in future to more than 2 data fields, although at the moment the feed creation function will only create one or two data-fields.

**Comment by MarsFlyer** [comment]
Sounds good. I'll pull the latest version tomorrow and have a look.
P.S. How do you get the multivis working?

**Comment by TrystanLea** [comment]
I have merged with the master and updated the version on vis.openenergymonitor. There seems to be a small problem when the histogram feed is first created as the first value is really high 60000 or 70000, after the first value the next ones seem fine, I might see if I can see why later, Thanks a lot for adding the functionality!

**Pull request merged**